// tshash/index.js
"use strict";
var utils_1 = require('./utils');
var SHA1_1 = require('./SHA1');
exports.SHA1 = SHA1_1.default;
// For compatibility with nodemod...
var CRC32_1 = require('./CRC32');
exports.CRC32 = CRC32_1.default;
var uuids = require('./uuids');
exports.uuids = uuids;
var utils_2 = require('./utils');
exports.utf8Encode = utils_2.utf8Encode;
exports.utf8Decode = utils_2.utf8Decode;
exports.base32Encode = utils_2.base32Encode;
exports.base32Decode = utils_2.base32Decode;
exports.hexEncode = utils_2.hexEncode;
exports.hexDecode = utils_2.hexDecode;
function hash(data, hashFunc) {
    var u8 = utils_1.toUint8Array(data);
    var hashing = hashFunc.newHashing();
    hashing.update(u8);
    return hashing.digest();
}
exports.hash = hash;
/** @deprecated; just use hash(str, hashFunc) */
function hashString(str, hashFunc) {
    return hash(utils_1.utf8Encode(str), hashFunc);
}
exports.hashString = hashString;
function sha1Urn(thing) {
    return "urn:sha1:" + utils_1.base32Encode(hash(thing, SHA1_1.default));
}
exports.sha1Urn = sha1Urn;
