// tshash/InternetChecksum.js
"use strict";
// Based on code by James Hughes
// http://stackoverflow.com/questions/4113890/how-to-calculate-the-internet-checksum-from-a-byte-in-java
var InternetChecksumming = (function () {
    function InternetChecksumming() {
        this.sum = 0;
        this.phase = 0;
    }
    InternetChecksumming.prototype.reset = function () {
        this.sum = 0;
        this.phase = 0;
    };
    InternetChecksumming.prototype.update = function (buf) {
        var i = 0, length = buf.length, sum = this.sum, phase = this.phase;
        while (length > 0) {
            if (phase == 0) {
                sum += buf[i++] << 8;
                --length;
                phase = 1;
            }
            else {
                sum += buf[i++];
                --length;
                phase = 0;
            }
        }
        this.sum = sum;
        this.phase = phase;
    };
    InternetChecksumming.prototype.digestAsUint16 = function () {
        return (~((this.sum & 0xFFFF) + (this.sum >> 16))) & 0xFFFF;
    };
    InternetChecksumming.prototype.digest = function () {
        var a = new Uint8Array(2);
        new DataView(a.buffer).setUint16(0, this.digestAsUint16());
        return a;
    };
    return InternetChecksumming;
}());
exports.InternetChecksumming = InternetChecksumming;
var InternetChecksum = (function () {
    function InternetChecksum() {
    }
    InternetChecksum.newHashing = function () {
        return new InternetChecksumming();
    };
    return InternetChecksum;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = InternetChecksum;
